In [ ]:
# Vettorizzazione con TF-IDF con e senza stopwords


import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer

# Carico il dataset pulito
df = pd.read_csv("datasetpronto.csv")

# Estraggo la lista dei testi (plot puliti) e delle etichette (generi)
testi = df['Cleaned_Plot'].tolist()
etichette = df['Genre'].tolist()

# Versione CON le stopwords (come nel tuo codice originale)
vectorizer = TfidfVectorizer()
X = vectorizer.fit_transform(testi)

print("Vettorizzazione CON stopwords")
print(f"Numero di testi vettorizzati: {X.shape[0]}")
print(f"Numero di caratteristiche (vocaboli unici): {X.shape[1]}")
print()

# Versione SENZA le stopwords
vectorizer_no_stopwords = TfidfVectorizer(stop_words='english')
X_no_stopwords = vectorizer_no_stopwords.fit_transform(testi)

print("Vettorizzazione SENZA stopwords")
print(f"Numero di testi vettorizzati: {X_no_stopwords.shape[0]}")
print(f"Numero di caratteristiche (vocaboli unici): {X_no_stopwords.shape[1]}")

Vettorizzazione CON stopwords
Numero di testi vettorizzati: 4398
Numero di caratteristiche (vocaboli unici): 57822

Vettorizzazione SENZA stopwords
Numero di testi vettorizzati: 4398
Numero di caratteristiche (vocaboli unici): 57519


list

Il numero di caratteristiche senza le stopword è minore perché semplicemente ha tolto 303 stopwords (303 parole in meno!)

In [ ]:
import numpy as np

plot_index = 84  # indice del plot. Se io cambio 84 e metto un numero a caso, mi dice i valori tf-idf delle parole del plot numero ad es. 37
feature_array = np.array(vectorizer.get_feature_names_out())
tfidf_scores = X[plot_index].toarray().flatten()

# Ottengo e ordino le parole con i pesi più alti
top_n = 20
top_indices = tfidf_scores.argsort()[-top_n:][::-1]
top_words = feature_array[top_indices]
top_weights = tfidf_scores[top_indices]

# Stampa le top parole
for word, weight in zip(top_words, top_weights):
    print(f"{word}: {weight:.4f}")

fedayeen: 0.2972
sortie: 0.2833
insurgents: 0.2659
foe: 0.2498
quarter: 0.2498
iraq: 0.2389
routine: 0.2004
survival: 0.1908
draw: 0.1865
patrol: 0.1798
superior: 0.1787
ambushed: 0.1767
fast: 0.1762
combat: 0.1720
skills: 0.1649
knowledge: 0.1590
action: 0.1543
desperate: 0.1527
training: 0.1522
soldiers: 0.1513


In [ ]:
#Vettorizzazione con CountVectorizer con e senza stopwords

import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer

# Carico il dataset pulito
df = pd.read_csv("datasetpronto.csv")

# Estraggo la lista dei testi e delle etichette
testi = df['Cleaned_Plot'].tolist()
etichette = df['Genre'].tolist()

# Creo un vettorizzatore CountVectorizer (con stopwords inglesi)
CVvectorizer = CountVectorizer()

# Applico il vettorizzatore ai testi
cvX = CVvectorizer.fit_transform(testi)

print("Vettorizzazione CON stopwords")
print(f"Numero di testi vettorizzati: {cvX.shape[0]}")
print(f"Numero di caratteristiche (vocaboli unici): {cvX.shape[1]}")
print()

# --- Versione SENZA le stopwords ---
CVvectorizer_no_stopwords = CountVectorizer(stop_words='english')
cvX_no_stopwords = CVvectorizer_no_stopwords.fit_transform(testi)

print("Vettorizzazione SENZA stopwords")
print(f"Numero di testi vettorizzati: {cvX_no_stopwords.shape[0]}")
print(f"Numero di caratteristiche (vocaboli unici): {cvX_no_stopwords.shape[1]}")

✅ Vettorizzazione CON stopwords
Numero di testi vettorizzati: 4398
Numero di caratteristiche (vocaboli unici): 57822

✅ Vettorizzazione SENZA stopwords
Numero di testi vettorizzati: 4398
Numero di caratteristiche (vocaboli unici): 57519


D'ORA IN POI INIZIA IL MODELLO.

In [ ]:
# importo GridSearchCV
from sklearn.model_selection import GridSearchCV




# importo classificatore SVC
from sklearn.svm import SVC




# instantiate classifier with default hyperparameters with kernel=rbf, C=1.0 and gamma=auto
svc=SVC()




parameters = [
    {'C': [1, 10, 100], 'kernel': ['linear']},
    {'C': [1, 10, 100], 'kernel': ['rbf']},
    {'C': [1, 10, 100], 'kernel': ['poly']}
]








grid_search = GridSearchCV(estimator = svc,  
                           param_grid = parameters,
                           scoring = 'accuracy',
                           cv = 5,
                           verbose=1)




grid_search.fit(X, etichette)

In [4]:
# esamino il miglior modello




# best score achieved during the GridSearchCV
print('GridSearch CV best score : {:.4f}\n\n'.format(grid_search.best_score_))




# print parameters that give the best results
print('Parameters that give the best results :','\n\n', (grid_search.best_params_))




# print estimator that was chosen by the GridSearch
print('\n\nEstimator that was chosen by the search :','\n\n', (grid_search.best_estimator_))

NameError: name 'grid_search' is not defined

In [ ]:
#Addestramento modello con matrice IF-IDF con stopwords

from sklearn.model_selection import train_test_split


# Divisione 80% training, 20% test
X_train, X_test, y_train, y_test = train_test_split(X, etichette, test_size=0.2, random_state=42, stratify=etichette)


from sklearn.svm import SVC


# Istanzia il modello con i parametri migliori trovati
best_model = SVC(kernel='linear', C=1)


# Addestramento
best_model.fit(X_train, y_train)


from sklearn.metrics import classification_report, accuracy_score


# Predizioni
y_pred = best_model.predict(X_test)


# Report di valutazione
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

Accuracy: 0.7147727272727272

Classification Report:
               precision    recall  f1-score   support

      action       0.79      0.79      0.79       220
      comedy       0.63      0.67      0.65       220
       drama       0.61      0.60      0.60       220
      horror       0.85      0.80      0.83       220

    accuracy                           0.71       880
   macro avg       0.72      0.71      0.72       880
weighted avg       0.72      0.71      0.72       880



In [27]:
#Addestramento modello con matrice IF-IDF senza stopwords

from sklearn.model_selection import train_test_split


# Divisione 80% training, 20% test
X_train, X_test, y_train, y_test = train_test_split(X_no_stopwords, etichette, test_size=0.2, random_state=42, stratify=etichette)


from sklearn.svm import SVC


# Istanzia il modello con i parametri migliori trovati
best_model = SVC(kernel='linear', C=1)


# Addestramento
best_model.fit(X_train, y_train)


from sklearn.metrics import classification_report, accuracy_score


# Predizioni
y_pred = best_model.predict(X_test)


# Report di valutazione
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

Accuracy: 0.7238636363636364

Classification Report:
               precision    recall  f1-score   support

      action       0.78      0.80      0.79       220
      comedy       0.64      0.69      0.66       220
       drama       0.62      0.62      0.62       220
      horror       0.86      0.79      0.82       220

    accuracy                           0.72       880
   macro avg       0.73      0.72      0.73       880
weighted avg       0.73      0.72      0.73       880



In [ ]:
#PROVE PER FEATURES 1/2

import numpy as np

# Recupera il vocabolario usato nel modello TF-IDF
feature_names = vectorizer_no_stopwords.get_feature_names_out()

# Ottieni i coefficienti del classificatore lineare
# Ogni riga di coef_ corrisponde a una classe
classi = best_model.classes_  # ['action', 'comedy', 'drama', 'horror']
coefficienti = best_model.coef_

# Trova l'indice associato alla classe 'horror'
indice_horror = list(classi).index('horror')

# Ottieni i coefficienti per la classe 'horror'
coefficienti_horror = coefficienti[indice_horror].toarray().flatten()

# Ordina le features per importanza (in valore assoluto o positivo/negativo)
top_n = 20  # quante parole vuoi visualizzare
top_indici = np.argsort(coefficienti_horror)[-top_n:][::-1]

print(f"Top {top_n} parole più indicative per la classe 'horror':")
for indice in top_indici:
    print(f"{feature_names[indice]}: {coefficienti_horror[indice]:.4f}")

Top 20 parole più indicative per la classe 'horror':
wedding: 1.4484
real: 1.3587
hotel: 1.3180
comedy: 1.2933
local: 1.2604
lady: 1.2511
getting: 1.2491
lucy: 1.2149
girlfriend: 1.1995
betty: 1.1953
trip: 1.1879
attractive: 1.1836
dog: 1.1427
gay: 1.1334
andrew: 1.1293
army: 1.1181
kate: 1.1021
accidentally: 1.0948
date: 1.0786
group: 1.0767


In [ ]:
#PROVE PER FEATURES 2/2

import numpy as np

# Recupera il vocabolario
feature_names = vectorizer_no_stopwords.get_feature_names_out()

# Ottieni i coefficienti per tutte le classi
coefficienti = best_model.coef_.toarray()
classi = best_model.classes_

# Ottieni i pesi della classe horror
indice_horror = list(classi).index('horror')
pesi_horror = coefficienti[indice_horror]

# Calcola la media dei pesi delle ALTRE classi
pesi_altre_classi = np.mean(np.delete(coefficienti, indice_horror, axis=0), axis=0)

# Calcola la differenza dei pesi: horror vs altri
delta = pesi_horror - pesi_altre_classi

# Ordina in base alla differenza
top_n = 20
indici_top = np.argsort(delta)[-top_n:][::-1]

print(f"Top {top_n} parole distintive per 'horror' rispetto alle altre classi:")
for indice in indici_top:
    print(f"{feature_names[indice]}: delta={delta[indice]:.4f}, horror_weight={pesi_horror[indice]:.4f}, others_mean={pesi_altre_classi[indice]:.4f}")

Top 20 parole distintive per 'horror' rispetto alle altre classi:
house: delta=1.7066, horror_weight=0.3272, others_mean=-1.3794
dr: delta=1.6870, horror_weight=0.1189, others_mean=-1.5680
kate: delta=1.5128, horror_weight=1.1021, others_mean=-0.4107
group: delta=1.4982, horror_weight=1.0767, others_mean=-0.4215
accidentally: delta=1.4601, horror_weight=1.0948, others_mean=-0.3654
betty: delta=1.4583, horror_weight=1.1953, others_mean=-0.2629
lucy: delta=1.4264, horror_weight=1.2149, others_mean=-0.2116
comedy: delta=1.3855, horror_weight=1.2933, others_mean=-0.0922
trip: delta=1.3815, horror_weight=1.1879, others_mean=-0.1936
hotel: delta=1.3777, horror_weight=1.3180, others_mean=-0.0596
girls: delta=1.3653, horror_weight=1.0452, others_mean=-0.3201
real: delta=1.3623, horror_weight=1.3587, others_mean=-0.0037
lou: delta=1.3320, horror_weight=0.8916, others_mean=-0.4404
wedding: delta=1.3319, horror_weight=1.4484, others_mean=0.1164
girlfriend: delta=1.3079, horror_weight=1.1995, othe

In [30]:
#modello adesstrato con la matrice creata con CountVectorizer con stopwords

from sklearn.model_selection import train_test_split


# Divisione 80% training, 20% test
X_train, X_test, y_train, y_test = train_test_split(cvX, etichette, test_size=0.2, random_state=42, stratify=etichette)


from sklearn.svm import SVC


# Istanzia il modello con i parametri migliori trovati
best_model = SVC(kernel='linear', C=1)


# Addestramento
best_model.fit(X_train, y_train)


from sklearn.metrics import classification_report, accuracy_score


# Predizioni
y_pred = best_model.predict(X_test)


# Report di valutazione
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

Accuracy: 0.6136363636363636

Classification Report:
               precision    recall  f1-score   support

      action       0.69      0.64      0.67       220
      comedy       0.51      0.55      0.53       220
       drama       0.51      0.50      0.51       220
      horror       0.76      0.75      0.76       220

    accuracy                           0.61       880
   macro avg       0.62      0.61      0.61       880
weighted avg       0.62      0.61      0.61       880



In [17]:
#modello adesstrato con la matrice creata con CountVectorizer senza stopwords

from sklearn.model_selection import train_test_split


# Divisione 80% training, 20% test
X_train, X_test, y_train, y_test = train_test_split(cvX_no_stopwords, etichette, test_size=0.2, random_state=42, stratify=etichette)


from sklearn.svm import SVC


# Istanzia il modello con i parametri migliori trovati
best_model = SVC(kernel='linear', C=1)


# Addestramento
best_model.fit(X_train, y_train)


from sklearn.metrics import classification_report, accuracy_score


# Predizioni
y_pred = best_model.predict(X_test)


# Report di valutazione
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

Accuracy: 0.6318181818181818

Classification Report:
               precision    recall  f1-score   support

      action       0.69      0.66      0.67       220
      comedy       0.54      0.58      0.56       220
       drama       0.54      0.55      0.55       220
      horror       0.78      0.73      0.76       220

    accuracy                           0.63       880
   macro avg       0.64      0.63      0.63       880
weighted avg       0.64      0.63      0.63       880



D'ORA IN POI FACCIAMO TEST SUL MODELLO CON IL DATASET SPORCO.

In [20]:
# Vettorizzazione con TF-IDF con e senza stopwords CON DATASET SPORCO


import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer

# Carico il dataset pulito
df = pd.read_csv("datasetsporcopronto.csv")

# Estraggo la lista dei testi (plot puliti) e delle etichette (generi)
testi = df['Plot'].tolist()
etichette = df['Genre'].tolist()

# --- Versione CON le stopwords (come nel tuo codice originale) ---
vectorizer = TfidfVectorizer()
Xsporca = vectorizer.fit_transform(testi)

print("Vettorizzazione CON stopwords")
print(f"Numero di testi vettorizzati: {Xsporca.shape[0]}")
print(f"Numero di caratteristiche (vocaboli unici): {Xsporca.shape[1]}")
print()

# --- Versione SENZA le stopwords ---
vectorizer_no_stopwords = TfidfVectorizer(stop_words='english')
Xsporca_no_stopwords = vectorizer_no_stopwords.fit_transform(testi)

print("Vettorizzazione SENZA stopwords")
print(f"Numero di testi vettorizzati: {Xsporca_no_stopwords.shape[0]}")
print(f"Numero di caratteristiche (vocaboli unici): {Xsporca_no_stopwords.shape[1]}")

Vettorizzazione CON stopwords
Numero di testi vettorizzati: 4398
Numero di caratteristiche (vocaboli unici): 49652

Vettorizzazione SENZA stopwords
Numero di testi vettorizzati: 4398
Numero di caratteristiche (vocaboli unici): 49353


In [ ]:
#Addestramento modello con matrice IF-IDF senza stopwords CON DATASET SPORCO (n.b.: perché riguardo al dataset sporco, vediamo solo senza stopwords)

from sklearn.model_selection import train_test_split


# Divisione 80% training, 20% test
X_train, X_test, y_train, y_test = train_test_split(Xsporca_no_stopwords, etichette, test_size=0.2, random_state=42, stratify=etichette)


from sklearn.svm import SVC


# Istanzia il modello con i parametri migliori trovati
best_model = SVC(kernel='linear', C=1)


# Addestramento
best_model.fit(X_train, y_train)


from sklearn.metrics import classification_report, accuracy_score


# Predizioni
y_pred = best_model.predict(X_test)


# Report di valutazione
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

Accuracy: 0.7170454545454545

Classification Report:
               precision    recall  f1-score   support

      action       0.77      0.80      0.78       220
      comedy       0.63      0.66      0.65       220
       drama       0.62      0.61      0.62       220
      horror       0.86      0.79      0.82       220

    accuracy                           0.72       880
   macro avg       0.72      0.72      0.72       880
weighted avg       0.72      0.72      0.72       880



DALLA NOSTRA RICERCA ABBIAMO VISTO CHE:
- L'ACCURACY CON TF-IDF VECTORIZER E' MAGGIORE RISPETTO A COUNTVECTORIZER.

TF-IDF
- L'ACCURACY CON TF-IDF VECTORIZER E' MAGGIORE QUANDO RIMUOVO LE STOPWORDS RISPETTO A QUANDO LE TENGO.

COUNTVECTORIZER
- L'ACCURACY CON COUNTVECTORIZER E' MAGGIORE QUANDO RIMUOVO LE STOPWORDS RISPETTO A QUANDO LE TENGO.

DATASET SPORCO
- TF-IDF / senza stopwords / DATASET PULITO          CONFRONTO CON          TF-IDF / senza stopwords / DATASET SPORCO
         Accuracy: 0.7238636363636364                                                Accuracy: 0.7170454545454545

Quindi, pulire il dataset porta ad una maggiore accuracy.


